lib/src/source/hosted.dart

```diff
@@ -537,6 +537,62 @@ class HostedSource extends CachedSource {
     return result;
   }
 
+  Future<_PackageAdvisories> _fetchAdvisories(
+    _RefAndCache refAndCache,
+  ) async {
+    final ref = refAndCache.ref;
+    final description = ref.description;
+    final cache = refAndCache.cache;
+
+    if (description is! HostedDescription) {
+      throw ArgumentError('Wrong source');
+    }
+
+    final packageName = description.packageName;
+    final hostedUrl = description.url;
+
+    final url = _listAdvisoriesUrl(ref);
+    log.io('Get security advisories from $url.');
+
+    final String bodyText;
+    final Map<String, dynamic> body;
+    final _PackageAdvisories result;
+    try {
+      bodyText = await withAuthenticatedClient(cache, Uri.parse(hostedUrl),
+          (client) async {
+        return await retryForHttp(
+            'fetching advisories for "$packageName" from "$url"', () async {
+          final request = http.Request('GET', url);
+          request.attachPubApiHeaders();
+          request.attachMetadataHeaders();
+          final response = await client.fetch(request);
+          return response.body;
+        });
+      });
+      final decoded = jsonDecode(bodyText);
+      if (decoded is! Map<String, dynamic>) {
+        throw FormatException('security advisories must be a mapping');
+      }
+      body = decoded;
+
+      result = _PackageAdvisories(
+        body['advisories'] as List,
```

Comment:
We cannot do a type-cast here. We have to validate this is actually a list and throw if not